// ngx-http-client/dist/src/params.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
* A `HttpParameterCodec` that uses `encodeURIComponent` and `decodeURIComponent` to
* serialize and parse URL parameter keys and values.
*
* @stable
*/
var WebHttpUrlEncodingCodec = /** @class */ (function () {
    function WebHttpUrlEncodingCodec() {
    }
    WebHttpUrlEncodingCodec.prototype.encodeKey = function (k) { return standardEncoding(k); };
    WebHttpUrlEncodingCodec.prototype.encodeValue = function (v) { return standardEncoding(v); };
    WebHttpUrlEncodingCodec.prototype.decodeKey = function (k) { return decodeURIComponent(k); };
    WebHttpUrlEncodingCodec.prototype.decodeValue = function (v) { return decodeURIComponent(v); };
    return WebHttpUrlEncodingCodec;
}());
export { WebHttpUrlEncodingCodec };
function standardEncoding(v) {
    return encodeURIComponent(v);
    // .replace(/%40/gi, '@')
    // .replace(/%3A/gi, ':')
    // .replace(/%24/gi, '$')
    // .replace(/%2C/gi, ',')
    // .replace(/%3B/gi, ';')
    // .replace(/%2B/gi, '+')
    // .replace(/%3D/gi, '=')
    // .replace(/%3F/gi, '?')
    // .replace(/%2F/gi, '/');
}
