// ngx-http-client/dist/test/jsonp_mock.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var MockScriptElement = /** @class */ (function () {
    function MockScriptElement() {
        this.listeners = {};
    }
    MockScriptElement.prototype.addEventListener = function (event, handler) {
        this.listeners[event] = handler;
    };
    MockScriptElement.prototype.removeEventListener = function (event) { delete this.listeners[event]; };
    return MockScriptElement;
}());
export { MockScriptElement };
var MockDocument = /** @class */ (function () {
    function MockDocument() {
    }
    MockDocument.prototype.createElement = function (tag) {
        return new MockScriptElement();
    };
    Object.defineProperty(MockDocument.prototype, "body", {
        get: function () { return this; },
        enumerable: true,
        configurable: true
    });
    MockDocument.prototype.appendChild = function (node) { this.mock = node; };
    MockDocument.prototype.removeNode = function (node) {
        if (this.mock === node) {
            this.mock = null;
        }
    };
    MockDocument.prototype.mockLoad = function () { this.mock.listeners.load(null); };
    MockDocument.prototype.mockError = function (err) { this.mock.listeners.error(err); };
    return MockDocument;
}());
export { MockDocument };
